.github/workflows/build-test-osx-x86.yaml

```diff
@@ -1,29 +1,56 @@
 # This workflow builds and tests the semgrep-core binary for macOS X86
 # and generates osx-wheel for pypi.
 
+# This workflow will cache the ~/.opam directory, which is not symmetic with our other architecture's build processes.
+# This runs on GHA-hosted runners, so without caching, it runs very slowly. The Linux build uses a container to bring in the
+# required dependencies, and the M1 build runs on a self-hosted machine where caching isn't necessary.
+# To update to a new version of OCaml, we can switch out `OPAM_SWITCH_NAME` var which will update the cache key, and lead to a cache miss
+# on the new builds.
+
 # coupling: if you modify this file, modify also build-test-osx-m1.yaml
 name: build-test-osx-x86
 
 on:
   workflow_dispatch:
+    inputs:
+      use-cache:
+        description: "Use Opam Cache - uncheck the box to disable use of the opam cache, meaning a long-running but completely from-scratch build."
+        required: true
+        type: boolean
+        default: true
   workflow_call:
+    inputs:
+      use-cache:
+        description: "Use Opam Cache - uncheck the box to disable use of the opam cache, meaning a long-running but completely from-scratch build."
+        required: false
+        type: boolean
+        default: true
 
 jobs:
   build-core-osx:
     name: Build the OSX binaries
-    runs-on: [self-hosted, macOS, X64]
+    runs-on: macos-12
+    env:
+      # This name is used in the cache key. If we update to a newer version of ocaml, we'll want to change the OPAM_SWITCH_NAME
+      # as well to avoid issues with caching.
+      OPAM_SWITCH_NAME: "4.14.0"
     steps:
-      # see the comment in build-test-osx-m1.yaml why we need this step
-      - name: Clean Checkout
-        run: |
-          rm -rf ./*
-          rm -f .gitmodules
-          rm -rf .git
       - uses: actions/checkout@v3
         with:
           submodules: true
-      - run: |
-          ./scripts/osx-setup-for-release.sh
+      # Note that this action handles both cache read and cache write.
```

Comment:
Ok I still don't understand how this caching works.
When do we save this ~/.opam directory? At the very end of the jobs? or just after the next step?
Because we may want to cache only the content of ~/.opam after scripts/osx-setup-for-release.sh
has finished (where it will run the opam switch create 4.14.0 the very first time), but not the content
of ~/.opam after 'opam exec -- make core', which does extra thing.
Is it possible to give a scope to the cache? To tell GHA to save the content of a directory only after a certain step?

Reply:
@aryx yes, this step runs at the end of the job. It is possible to break this into two steps - the cache read and cache write individually - I can do that if we'd prefer and it would make things more clear. 

I have some follow up work here, I can include this as well.